### PR TITLE
Initial validating XML encoder

### PR DESCRIPTION
**Pull request by deviantintegral**
- [x] Wire up this repository against packagist
- [x] Wire up this repository against CircleCI
- [x] Tests? Or, marking the existing methods as incomplete with docs.

### CODE REVIEW

File: src/Encoder.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+        if (!file_exists($xsd_path) && !is_readable($xsd_path)) {
+            throw new \InvalidArgumentException('The XSD could not be read');
+        }
+
+        // We load the XSD into memory so we avoid having to enable external
+        // XML entities and the security implications of that.
+        $this->xsd = file_get_contents($xsd_path);
+
+        if (!$this->xsd) {
```
Comment (deviantintegral) [inline src/Encoder.php]: I should inline this with line 39.

File: src/Encoder.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        $dom = $this->createDomDocument($context);
+        $dom->loadXML($xml, LIBXML_NOENT | LIBXML_NOBLANKS);
+        $valid = $dom->schemaValidateSource($this->xsd);
+
+        $error = libxml_get_last_error();
+        libxml_use_internal_errors($internalErrors);
+        libxml_disable_entity_loader($disableEntities);
+
+        if (!$valid) {
+            // @todo how do we format xml responses?
```
Comment (deviantintegral) [inline src/Encoder.php]: This is probably a TODO for callers of this code, not this library.

File: src/Encoder.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            throw new XsdValidationException($error);
+        }
+
+        return $dom->saveXML();
+    }
+
+    /**
+     * Create a DOM document, taking serializer options into account.
```
Comment (deviantintegral) [inline src/Encoder.php]: I should note this is copy-pasta from the parent class.

File: src/Encoder.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    public function __construct($rootNodeName, $xsd_path)
+    {
+        parent::__construct($rootNodeName);
+
+        if (!file_exists($xsd_path) && !is_readable($xsd_path)) {
+            throw new \InvalidArgumentException('The XSD could not be read');
+        }
+
+        // We load the XSD into memory so we avoid having to enable external
+        // XML entities and the security implications of that.
```
Comment (m4olivei) [inline src/Encoder.php]: Just out of curiosity. I don't understand, can you elaborate on:

> so we avoid having to enable external XML entities and the security implications of that

File: src/Encoder.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+     * Create a DOM document, taking serializer options into account.
+     *
+     * @param array $context options that the encoder has access to
+     *
+     * @return \DOMDocument
+     */
+    private function createDomDocument(array $context)
```
Comment (m4olivei) [inline src/Encoder.php]: Booo private method, common Symfony. Too bad not parent method isn't protected.

Comment (deviantintegral) [inline src/Encoder.php]: Yeah, the whole reason I reparse the XML in `encode()` is that the methods are all private.

File: src/Encoder.php (hunk 4 above)
Comment (deviantintegral) [inline src/Encoder.php]: https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Processing has some good examples.

**Pull request merged**